### PR TITLE
[IMP] project , _*: improve portal list view and website form

### PR DESCRIPTION
**Pull request by shap-odoo**
_* = website_form_project

- removed the color code to the portal list view for task stage display as the fact that it is folded or not no longer determines whether a task is closed or open.

- added some additional fields for capturing customer contact details.
- modified the view page after submitting task so that user can follow up on the progress of the task.
- optimize the Submit button for dynamic behavior based on user actions.
- Improved the website form based on user actions.

  task-3635784

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website_project/views/project_portal_project_task_template.xml [hunk 1]
```diff
@@ -1,5 +1,41 @@
...
+                            <a class="btn btn-primary" t-if="task and request.session.uid and task.project_id"  t-attf-href="/my/task/#{task.id}">View Task</a>
+                            <a class="btn btn-primary" href='/'>Go to the Homepage</a>
+                        </div>
+                    </div>
+                </div>
+            </div>
+        </t>
+    </template>
+
+    <record id="aboutus_page" model="website.page">
```
Comment (pkyriakou) [inline addons/website_project/views/project_portal_project_task_template.xml:32]: I'd change the id here to something more relevant

**Review by pkyriakou — COMMENTED** [review]
Thank you @shap-odoo :100: 
I left some comments for you :grin:

File: addons/website_project/controllers/main.py [hunk 2]
```diff
@@ -12,8 +12,15 @@ def insert_record(self, request, model, values, custom, meta=None):
...
-            if visitor_partner:
-                values['partner_id'] = visitor_partner.id
+            if not visitor_partner:
+                visitor_partner = visitor_partner.sudo().create({
+                    'email': request.params.get('email_from', False),
+                    'name': request.params.get('partner_name', False),
+                    'phone': request.params.get('partner_phone', False),
+                    'company_name': request.params.get('partner_company_name', False),
+                    'lang': request.lang.code,
+                })
```
Comment (ppr-odoo) [inline addons/website_project/controllers/main.py:22]:
```suggestion
                def _prepare_visitor_partner_vals(self):
                  return {
                      'email': request.params.get('email_from', False),
                      'name': request.params.get('partner_name', False),
                      'phone': request.params.get('partner_phone', False),
                      'company_name': request.params.get('partner_company_name', False),
                      'lang': request.lang.code,
                  }
                  visitor_partner_vals = .._prepare_visitor_partner_vals()
                   values['partner_id'] = visitor_partner.sudo().create(visitor_partner_vals).id
```

code is not tested and create separate vals method so anyone can easily customization.

File: addons/website_project/static/src/js/options.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    /**
+     * @override
+     */
+    async _renderCustomXML(uiFragment) {
+        await this._super(uiFragment);
+        const active_model=this.$target[0].dataset.model_name
+        let element = this.$target[0].querySelector(".s_website_form_send");
+        if (active_model=='project.task') {
+            element.innerText = 'Submit Task';
+        }
```
Comment (ppr-odoo) [inline addons/website_project/static/src/js/options.js:15]:
```suggestion
        await this._super(...arguments);
        const activeModel = this.$target[0].dataset.model_name;
        if (active_model === 'project.task') {
           this.$target[0].querySelector(".s_website_form_send").innerText = _t('Submit Task');
        }
```

code is not tested.thanks

**Review by pkyriakou — COMMENTED** [review]
Thank you @shap-odoo :100: 
LGTM :green_circle: :v:

**Review by auon-odoo — APPROVED** [review]

File: addons/website_project/static/src/js/options.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+options.registry.WebsiteFormEditor = options.registry.WebsiteFormEditor.extend({
+    /**
+     * @override
+     */
+    async _renderCustomXML(uiFragment) {
+        await this._super(...arguments);
+        const activeModel = this.$target[0].dataset.model_name;
+        if (activeModel === 'project.task') {
+            this.$target[0].querySelector(".s_website_form_send").innerText = _t('Submit Task');
+        }
```
Comment (auon-odoo) [inline addons/website_project/static/src/js/options.js:15]: Why is that needed?

**Review by auon-odoo — CHANGES_REQUESTED** [review]

File: addons/website_project/static/src/js/options.js (hunk 4 above)
Comment (auon-odoo) [reply addons/website_project/static/src/js/options.js:15]: We can also drop that :)

**Review by auon-odoo — APPROVED** [review]
LGTM, thanks!

File: addons/project/models/project_task.py [hunk 5]
```diff
@@ -205,6 +205,10 @@ def _read_group_personal_stage_type_ids(self, stages, domain, order):
     partner_id = fields.Many2one('res.partner',
         string='Customer', recursive=True, tracking=True, compute='_compute_partner_id', store=True, readonly=False,
         domain="['|', ('company_id', '=?', company_id), ('company_id', '=', False)]", )
+    # Used to submit tasks from a contact form
+    partner_name = fields.Char(string='Customer Name', related="partner_id.name", store=True, readonly=False)
+    partner_phone = fields.Char(string='Customer Phone', related="partner_id.phone", store=True, readonly=False)
+    partner_company_name = fields.Char(string='Company Name', related="partner_id.company_name", store=True, readonly=False)
```
Comment (xavierbol) [inline addons/project/models/project_task.py:211]: Since those fields are only useful when `website_project` module, why not adding them in `website_project` module?

File: addons/website_project/controllers/main.py [hunk 6]
```diff
@@ -36,3 +38,12 @@ def insert_record(self, request, model, values, custom, meta=None):
...
         return res
+
+    def _prepare_visitor_partner_vals(self, request):
+        return {
+            'email': request.params.get('email_from', False),
+            'name': request.params.get('partner_name', False),
+            'phone': request.params.get('partner_phone', False),
+            'company_name': request.params.get('partner_company_name', False),
+            'lang': request.lang.code,
+        }
```
Comment (xavierbol) [inline addons/website_project/controllers/main.py:49]: It does not seem, that hook is overriden somewhere, right? If yes, then the hook is useless, we could directly do the content of that method inside the method in which we call this one.

**Review by xavierbol — COMMENTED** [review]
Thanks for your work, I left some comments. 🙂

File: addons/website_project/controllers/main.py (hunk 6 above)
Comment (kais-odoo) [reply addons/website_project/controllers/main.py]: without_store


if we don't store these fields it will show those values in other information.

Comment (ppr-odoo) [reply addons/website_project/controllers/main.py]: > without_store
> 
> if we don't store these fields it will show those values in other information.

let me confirm with aju

Comment (kais-odoo) [reply addons/website_project/controllers/main.py]: my bad.. :see_no_evil:  no need to store it.

Comment (ppr-odoo) [reply addons/website_project/controllers/main.py]: > my bad.. 🙈 no need to store it.

Alright, we no longer need to confirm with the PO.

File: addons/website_project/views/project_portal_project_task_template.xml (hunk 1 above)
Comment (xavierbol) [reply addons/website_project/views/project_portal_project_task_template.xml:38]: I agree with @pkyriakou, we should give a better id since `About us` page is not really the content that we will display. 😅 

maybe `task_submitted_page`?

File: addons/website_project/models/project_task.py [hunk 7]
```diff
@@ -8,3 +8,7 @@ class ProjectTask(models.Model):
 
     # Need this field to check there is no email loops when Odoo reply automatically
     email_from = fields.Char('Email From')
+    # Used to submit tasks from a contact form
+    partner_name = fields.Char(string='Customer Name', related="partner_id.name", readonly=False)
+    partner_phone = fields.Char(string='Customer Phone', related="partner_id.phone", readonly=False)
+    partner_company_name = fields.Char(string='Company Name', related="partner_id.company_name", readonly=False)
```
Comment (xavierbol) [inline addons/website_project/models/project_task.py:14]: Are you sure we don't need to store them? if yes then okay let's merge that task. 🙂

Comment (kais-odoo) [reply addons/website_project/models/project_task.py:14]: yes, No need to store them. we can merge it.

File: addons/website_project/controllers/main.py [hunk 8]
```diff
@@ -31,6 +31,9 @@ def extract_data(self, model, values):
         if model.model == 'project.task' and values.get('email_from'):
             partners_list = request.env['mail.thread'].sudo()._mail_find_partner_from_emails([values['email_from']])
             partner_id = partners_list[0].id if partners_list else False
+            data['record']['partner_name'] = values['partner_name']
+            data['record']['partner_phone'] = values['partner_phone']
+            data['record']['partner_company_name'] = values['partner_company_name']
```
Comment (xavierbol) [inline addons/website_project/controllers/main.py:36]: If the partner exists, then we should just set those data only if the field is empty on the partner record.

What do you think? 

Otherwise, we will each time erase the data once the customer submit a task/ticket.

Comment (ppr-odoo) [reply addons/website_project/controllers/main.py]: > If the partner exists, then we should just set those data only if the field is empty on the partner record.
> 
> What do you think?
> 
> Otherwise, we will each time erase the data once the customer submit a task/ticket.

If the customer or user adds new data, such as an email, company name , we should store the new data in the task. Why should we use old data?

Comment (xavierbol) [reply addons/website_project/controllers/main.py]: Because it is untrusted data and so the customer or any user could easily change the partner data, each time he submits a ticket/task.

File: addons/website_project/controllers/main.py [hunk 9]
```diff
@@ -35,4 +35,14 @@ def extract_data(self, model, values):
             data['record']['email_from'] = values['email_from']
             if not partner_id:
                 data['record']['email_cc'] = values['email_from']
+                data['record']['partner_name'] = values['partner_phone']
```
Comment (xavierbol) [inline addons/website_project/controllers/main.py:38]:
```suggestion
                data['record']['partner_phone'] = values['partner_phone']
```

File: addons/website_project/controllers/main.py [hunk 10]
```diff
@@ -35,4 +35,14 @@ def extract_data(self, model, values):
...
+                data['record']['partner_name'] = values['partner_name']
+                data['record']['partner_company_name'] = values['partner_company_name']
+            else:
+                if not partners_list[0].phone:
+                    data['record']['partner_name'] = values['partner_phone']
+                if not partners_list[0].name:
+                    data['record']['partner_name'] = values['partner_name']
+                if not partners_list[0].company_name:
+                    data['record']['partner_company_name'] = values['partner_company_name']
```
Comment (xavierbol) [inline addons/website_project/controllers/main.py:47]: We could change a bit the code to get the `partner` recordset instead of each time doing `partners_list[0]...`

**Review by xavierbol — COMMENTED** [review]
Thanks for your work, I left some comments.

**Comment by ppr-odoo** [comment]
@xavierbol Now runbot happy.thanks

File: addons/website_project/controllers/main.py [hunk 11]
```diff
@@ -30,9 +30,19 @@ def extract_data(self, model, values):
         data = super().extract_data(model, values)
         if model.model == 'project.task' and values.get('email_from'):
             partners_list = request.env['mail.thread'].sudo()._mail_find_partner_from_emails([values['email_from']])
-            partner_id = partners_list[0].id if partners_list else False
-            data['record']['partner_id'] = partner_id
+            partner_id = partners_list[0] if partners_list else False
```
Comment (xavierbol) [inline addons/website_project/controllers/main.py:33]: We have to do that
```suggestion
            partner_id = partners_list[0] if partners_list else self.env['res.partner']
```

otherwise `partner_id.id` will crash since `False` has no attribute `id`...

I would also rename the variable

```suggestion
            partner = partners_list[0] if partners_list else self.env['res.partner']
```

File: addons/website_project/controllers/main.py [hunk 12]
```diff
@@ -30,9 +30,19 @@ def extract_data(self, model, values):
...
+                data['record']['partner_phone'] = values['partner_phone']
+                data['record']['partner_name'] = values['partner_name']
+                data['record']['partner_company_name'] = values['partner_company_name']
+            else:
+                if not partner_id.phone:
+                    data['record']['partner_phone'] = values['partner_phone']
+                if not partner_id.name:
+                    data['record']['partner_name'] = values['partner_name']
+                if not partner_id.company_name:
+                    data['record']['partner_company_name'] = values['partner_company_name']
```
Comment (xavierbol) [inline addons/website_project/controllers/main.py:47]:
```suggestion
            if partner:
                if not partner.phone:
                    data['record']['partner_phone'] = values['partner_phone']
                if not partner.name:
                    data['record']['partner_name'] = values['partner_name']
                if not partner.company_name:
                    data['record']['partner_company_name'] = values['partner_company_name']
            else:
                data['record']['email_cc'] = values['email_from']
                data['record']['partner_phone'] = values['partner_phone']
                data['record']['partner_name'] = values['partner_name']
                data['record']['partner_company_name'] = values['partner_company_name']
```

**Review by xavierbol — COMMENTED** [review]
Thanks again for your work, I left some comments. 🙂

**Comment by ppr-odoo** [comment]
@xavierbol When a partner is linked, we have to pass email_from in the record; otherwise, an error will occur. 
https://github.com/odoo/odoo/blob/master/addons/website/controllers/form.py#L272

Done other changes.

**Comment by xavierbol** [comment]
> @xavierbol When a partner is linked, we have to pass email_from in the record; otherwise, an error will occur. https://github.com/odoo/odoo/blob/master/addons/website/controllers/form.py#L272
> 
> Done other changes.

Can you add more details? Can you share me the error you got?

**Comment by ppr-odoo** [comment]
> > @xavierbol When a partner is linked, we have to pass email_from in the record; otherwise, an error will occur. https://github.com/odoo/odoo/blob/master/addons/website/controllers/form.py#L272
> > Done other changes.
> 
> Can you add more details? Can you share me the error you got?

I shared the error message in Discord.

File: addons/website_project/controllers/main.py [hunk 13]
```diff
@@ -30,9 +30,19 @@ def extract_data(self, model, values):
...
         if model.model == 'project.task' and values.get('email_from'):
             partners_list = request.env['mail.thread'].sudo()._mail_find_partner_from_emails([values['email_from']])
-            partner_id = partners_list[0].id if partners_list else False
-            data['record']['partner_id'] = partner_id
-            data['record']['email_from'] = values['email_from']
-            if not partner_id:
+            partner = partners_list[0] if partners_list else self.env['res.partner']
+            if partner:
+                data['record']['partner_id'] = partner.id
+                data['record']['email_from'] = values['email_from']
```
Comment (xavierbol) [inline addons/website_project/controllers/main.py:36]: Actually, before your changes `email_from` was given even if the partner is not found. 🤔 

Is it expected to only set it when we find a partner?

File: addons/website_project/views/project_portal_project_task_template.xml [hunk 14]
```diff
@@ -1,5 +1,47 @@
...
+                </div>
+            </div>
+        </t>
+    </template>
+
+    <record id="task_submitted_page" model="website.page">
+        <field name="is_published">True</field>
+        <field name="url">/your-task-has-been-submitted</field>
+        <field name="website_indexed" eval="False"/>
+        <field name="view_id" ref="task_submitted" />
```
Comment (xavierbol) [inline addons/website_project/views/project_portal_project_task_template.xml:42]: Nitpicking

```suggestion
        <field name="view_id" ref="task_submitted"/>
```

File: addons/website_project/controllers/main.py (hunk 13 above)
Comment (ppr-odoo) [reply addons/website_project/controllers/main.py]: > Actually, before your changes `email_from` was given even if the partner is not found. 🤔
> 
> Is it expected to only set it when we find a partner?

email_from is necessary for both partner and non-partner. I have made the changes.

FYI- We have to pass the  partner_id; otherwise, the task will have an empty customer field

thanks

File: addons/website_project/controllers/main.py [hunk 15]
```diff
@@ -30,9 +30,19 @@ def extract_data(self, model, values):
         data = super().extract_data(model, values)
         if model.model == 'project.task' and values.get('email_from'):
             partners_list = request.env['mail.thread'].sudo()._mail_find_partner_from_emails([values['email_from']])
-            partner_id = partners_list[0].id if partners_list else False
-            data['record']['partner_id'] = partner_id
+            partner = partners_list[0] if partners_list else self.env['res.partner']
             data['record']['email_from'] = values['email_from']
```
Comment (xavierbol) [inline addons/website_project/controllers/main.py:34]: We could also keep the other line to force partner_id to False otherwise an existing compute could set it for instance.

```suggestion
            data['record']['partner_id'] = partner.id
            data['record']['email_from'] = values['email_from']
```

File: addons/website_project/controllers/main.py [hunk 16]
```diff
@@ -30,9 +30,19 @@ def extract_data(self, model, values):
...
-            if not partner_id:
+            if partner:
+                data['record']['partner_id'] = partner.id
```
Comment (xavierbol) [inline addons/website_project/controllers/main.py:36]:
```suggestion
```

**Review by xavierbol — COMMENTED** [review]
Thanks for your work, I left 2 last comments, after that it will look good to me. 🙂

**Comment by ppr-odoo** [comment]
> Thanks for your work, I left 2 last comments, after that it will look good to me. 🙂

Done.thanks

**Review by xavierbol — APPROVED** [review]
LGTM, thanks for your work. 🙂 

robodoo r+
